internal/format/first_pass_visitor.go

```diff
@@ -0,0 +1,187 @@
+}
+
+func (v *firstPassVisitor) VisitSyntax(element *proto.Syntax) {
+	v.haveHitNonComment = true
+	if v.Syntax != nil {
+		v.AddFailure(element.Position, "duplicate syntax specified")
+		return
+	}
+	v.Syntax = element
+}
+
+func (v *firstPassVisitor) VisitPackage(element *proto.Package) {
+	v.haveHitNonComment = true
+	if v.Package != nil {
+		v.AddFailure(element.Position, "duplicate package specified")
+		return
+	}
+	v.Package = element
+}
+
+func (v *firstPassVisitor) VisitOption(element *proto.Option) {
+	// this will only hit file options since we don't do any
+	// visiting of children in this visitor
+	v.haveHitNonComment = true
+	if isProbablyCustomOption(element) {
+		v.ProbablyCustomOptions = append(v.ProbablyCustomOptions, element)
+	} else {
+		v.Options = append(v.Options, element)
+	}
+}
+
+func (v *firstPassVisitor) VisitImport(element *proto.Import) {
+	v.haveHitNonComment = true
+	v.Imports = append(v.Imports, element)
+}
+
+func (v *firstPassVisitor) VisitNormalField(element *proto.NormalField) {
+	v.haveHitNonComment = true
+}
+
+func (v *firstPassVisitor) VisitEnumField(element *proto.EnumField) {
+	v.haveHitNonComment = true
+}
+
+func (v *firstPassVisitor) VisitEnum(element *proto.Enum) {
+	v.haveHitNonComment = true
+}
+
+func (v *firstPassVisitor) VisitComment(element *proto.Comment) {
+	if !v.haveHitNonComment {
+		v.PComment(element)
+		v.P()
+	}
+}
+
+func (v *firstPassVisitor) VisitOneof(element *proto.Oneof) {
+	v.haveHitNonComment = true
+}
+
+func (v *firstPassVisitor) VisitOneofField(element *proto.OneOfField) {
+	v.haveHitNonComment = true
+}
+
+func (v *firstPassVisitor) VisitReserved(element *proto.Reserved) {
+	v.haveHitNonComment = true
+}
+
+func (v *firstPassVisitor) VisitRPC(element *proto.RPC) {
+	v.haveHitNonComment = true
+}
+
+func (v *firstPassVisitor) VisitMapField(element *proto.MapField) {
+	v.haveHitNonComment = true
+}
+
+func (v *firstPassVisitor) VisitGroup(element *proto.Group) {
+	v.haveHitNonComment = true
+}
+
+func (v *firstPassVisitor) VisitExtensions(element *proto.Extensions) {
+	v.haveHitNonComment = true
+}
+
+func (v *firstPassVisitor) PImports(imports []*proto.Import) {
+	if len(imports) == 0 {
+		return
+	}
+	sort.Slice(imports, func(i int, j int) bool { return imports[i].Filename < imports[j].Filename })
+	for _, i := range imports {
+		v.PComment(i.Comment)
+		// kind can be "weak", "public", or empty
+		// if weak or public, just print it out but with a space afterwards
+		// otherwise do not print anything
+		// https://developers.google.com/protocol-buffers/docs/reference/proto3-spec#import_statement
+		kind := i.Kind
+		if kind != "" {
+			kind = kind + " "
+		}
+		v.PWithInlineComment(i.InlineComment, `import `, kind, `"`, i.Filename, `";`)
+	}
+}
+
+func isProbablyCustomOption(option *proto.Option) bool {
+	// you can technically do ie google.protobuf.java_package
+	// but we're not going to handle this as I mean come on
+	return strings.HasPrefix(option.Name, "(")
+}
```

Comment:
Strange naming here - should we just call this `isCustomOption`? I don't like the "maybe" feeling :P

Reply:
I'm just going to get rid of the distinction and let sort take care of it.